### ISSUE TITLE
New way of making timelines outside of Dialogic.

### ISSUE THREAD

**Issue by coppolaemilio** [coppolaemilio/dialogic#580]
So, the timeline editor is one of my favourite features, but I've chatted and read a lot of feedback on what the writers need when making text heavy games, and they all want to use a regular text editor.

[Ink](https://www.inklestudios.com/ink/) seems to be a pretty popular way of writing this kind of games, and the most common by far is just good ol' [Ren'Py](https://www.renpy.org). Some even use google docs and copy-paste the text. Crazy, I know.

The rest of the people I spoke with were not using any kind of system and they were just writing it in the json/csv/script itself and not needing anything else.

I don't want to deprecate the timeline editor because I believe it is super nice to have, but instead of saving and loading that information into a json file that only a computer would want to read maybe we can come up with a way of storing that into a text file that is also human readable and editable.

I would love to keep the syntax as simple as possible without compromising the functionality of the events so I propose something like this:

```
[character join Max]
Max - Hello there! I'm Max and the only reason why I'm talking is because I have to fill out the screen with something
[audio play chill-music.ogg ]
[wait 2]
Max - Okay, bye!
```

A Text event could also be capable of parsing this kind of content, so if you want to write like this in the timeline editor you can do so. This will bring more flexibility in terms of how you write timelines specially for sections without a lot of special events.

At the moment this is a very early draft of the idea, but I wanted to write it out here so I can continue working on it and also getting feedback from anyone who might be interested.

**Comment by GimmiRuski** [comment]
What about making this the main way of creating timelines?

I imagine the main drawback (for new users) of having to create of a timeline through a regular text editor is the inability to know what events are available and how they are structured, so I'm proposing the following approach:

![d035ef80-8710-4b39-8e02-b622c6bda4ae](https://user-images.githubusercontent.com/44584436/137636981-257e545e-5598-45db-895a-37fb09f9a24b.png)

After clicking on an event button, you would get prompted about its parameters in a popup, which would then be used to add the corresponding line (or lines) to the timeline.

The main aim of this approach is to avoid having to maintain two timeline creation systems in parallel, while still maintaining some of the user-friendliness of the current GUI.

**Comment by bitbrain** [comment]
A very common tool is also [Yarn Editor](https://yarnspinner.dev/docs/writing/yarn-editor/#:~:text=The%20Yarn%20Editor%20is%20a,Yarn%20Editor%20in%20the%20browser.). Maybe introducing some sort of "plugin concept" into Dialogic could solve this problem. Dialogic itself could be kept clean and folks can just "plug in" what they need. Would https://github.com/coppolaemilio/dialogic/pull/644 solve that?

**Comment by alicerunsonfedora** [comment]
So glad I found this issue!

> Ink seems to be a pretty popular way of writing this kind of games, and the most common by far is just good ol' Ren'Py. Some even use google docs and copy-paste the text. Crazy, I know.

Yeah, I'm actually doing that right now for a project. Pretty painful 😣 

I have been considering making some form of an update to a Kotlin DSL package I made months ago to create the correct events and link them correctly, though I'm not sure how to go about doing it for the time being. 

Maybe if we had something like that so that Dialogic remains clean while having an optional language that can compile the text into JSON that Dialogic will understand? Languages like Swift and Kotlin support making DSLs via function builders, so I think it could work out...

**Comment by alicerunsonfedora** [comment]
So, a small update to this. I just managed to make a small utility in Swift to convert markdown documents to really basic Dialogic JSON. There isn't anything like signal emissions or scene changes, but for what I need as a starter tool, it gets the job done.

You can take a look at it at https://github.com/Indexing-Your-Heart/head-over-heels/tree/root/utils and see if there's anything in there that could help in making this alternate timeline script happen!

**Comment by Jowan-Spooner** [comment]
Update on this. This is implemented in the 2.0 branch, so we could in theory already close this issue. The question that remains is, if we will have a built-in dialogic text editor in godot.

**Comment by zaknafean** [comment]
Is it really on the table to ax the Dialogic UI as a whole?

**Comment by coppolaemilio** [comment]
We are not axing the UI! But now you can also edit timelines as plain text if you want to :)

**Comment by Jowan-Spooner** [comment]
> Is it really on the table to ax the Dialogic UI as a whole?

No I was considering having a separte "text edit" mode inside of godot. As some people prefer to write text, this would allow us to implement auto-completion or syntax higlighting for them in a script-editor like fashion. But for now the timeline editor will be the focus of deveolopement and text editing will have to happen outside of godot.

**Comment by Jowan-Spooner** [comment]
I will close this. If further features are wanted/needed, separate requests can be opened.
